### PR TITLE
Add Target::Pipe

### PR DESCRIPTION
**Pull request by TilCreator**
Now logs can be written to custom pipes and not just to stdout or stderr.

This is my take on #178.

I would be very happy to hear some feedback, since I'm quite unsure about a few changes:
- The [example](examples/custom_target.rs) seams overly complicated, mostly because I didn't want an example to write to a file. Maybe there is a better candidate than `std::sync::mpsc` used as a pipe for that?
- [`Builder::target_pipe`](src/fmt/writer/mod.rs#R110) is my idea for not breaking too much, but maybe there is a solution to implement that into [`Builder::target`](src/fmt/writer/mod.rs#R97) directly?
- In [`Buffer`](src/fmt/writer/termcolor/extern_impl.rs#R100) it is assumed that if `target_pipe` is `Some`, `test_target` is always `None`. Is there a better way to enforce that?

### CODE REVIEW

**Comment by TilCreator** [comment]
I can't reproduce the failed check locally and I also didn't change anything on the code that is failing. Plz help?

**Comment by jplatte** [comment]
It's `cargo clippy`, not `cargo check`, that's failing (and with a nightly version of clippy). I've pushed a commit to master that fixes that lint plus one more, rebasing should get rid of the CI failure now.

**Comment by TilCreator** [comment]
Ah, I wasn't using the nightly version of `clippy`. I rebased master and the checks succeed now. thx

File: src/fmt/writer/mod.rs
Comment (jplatte) [inline src/fmt/writer/mod.rs:26]:
```suggestion
    /// Logs will be sent to a custom pipe.
```

**Comment by jplatte** [comment]
You did the rebase the wrong way around (you should rebase your branch on top of master, not the other way around).

> [`Builder::target_pipe`](src/fmt/writer/mod.rs#R110) is my idea for not breaking too much, but maybe there is a solution to implement that into [`Builder::target`](src/fmt/writer/mod.rs#R97) directly?

What do you mean by "not breaking too much"? The derived traits for `Target`? I think I'd rather see those removed than there being a separate `target_pipe` builder method / field. I wonder why e.g. `Hash` is implemented for that in the first place.

**Comment by jplatte** [comment]
By the way, thank you for working on this!

**Comment by TilCreator** [comment]
Ups, I knew there was something wrong after at least the 5th git command, is it still worth trying to fix? ^^

**Comment by jplatte** [comment]
It should be as easy as

```sh
// if you haven't set up the remote yet, though I guess you have
git remote add upstream https://github.com/env-logger-rs/env_logger
git fetch upstream
// rebase your branch on top of upstream's master branch
git rebase upstream/master
```

even now. (git should detect the duplicate commit on your branch and get rid of it)

Followed by `git push --force` to tell git you really want to rewrite the history of the remote branch.

**Comment by TilCreator** [comment]
I have tried myself on keeping the pipe in the `Target::Pipe`, but [that](https://github.com/TilCreator/env_logger/commit/9e2f55a4e8be24fa88628f7b56d716151ee3f67e) is just such a mess, so many `Arc<Mutex<...>>`....
`Target` also gets used a lot after `build()` to check if stdout or stdout is used, so removing or "circumenting" the derived traits of `Target` is hard.

I will try something that takes the value out of `Target::Pipe` on `target()` or `build()` tomorrow.
Maybe I can even come up something better than `enum Target {..., Pipe(Option<Box<dyn io::Write + Send + 'static>>)}` after getting some sleep ^^

**Comment by TilCreator** [comment]
I'm quite happy with this version as it breaks very little if anything at all and it feels like a simple extension of the old `target` method.
Now I just need a better example as the [current one](examples/custom_target.rs) is just confusing and not super practical.

**Comment by SirWindfield** [comment]
I can take a closer look an Tuesday, up until then I don't have enough time for a full review :)

File: src/fmt/writer/termcolor/shim_impl.rs
Comment (SirWindfield) [inline src/fmt/writer/termcolor/shim_impl.rs:67]: You could merge the two (logic) branches into a single match arm:

````suggestion
pub(in crate::fmt::writer) fn print(&self, buf: &Buffer) -> io::Result<()> {
    // This impl uses the `eprint` and `print` macros
    // instead of using the streams directly.
    // This is so their output can be captured by `cargo test`.
    match self.target {
        // Safety: If the target type is `Pipe`, `target_pipe` will always be non-empty.
        TargetType::Pipe => self
            .target_pipe
            .as_ref()
            .unwrap()
            .lock()
            .unwrap()
            .write_all(&buf.0)?,
        TargetType::Stdout => print!("{}", String::from_utf8_lossy(&buf.0)),
        TargetType::Stderr => eprint!("{}", String::from_utf8_lossy(&buf.0)),
    }

    Ok(())
}
````

As far as I can tell the safety does apply to the PR.

File: examples/custom_target.rs
Comment (SirWindfield) [inline examples/custom_target.rs:19]: 2018 idoms, you should be able to just call `use log::{info, error, warn};`. Is there a policy in-place for examples and what edition they should target? If not, I would use the newer syntax.

File: examples/custom_target.rs
Comment (SirWindfield) [inline examples/custom_target.rs:52]:
```suggestion
        // Normally using a pipe as a target would mean a value of false, but this forces it to be true.
```

File: examples/custom_target.rs
Comment (SirWindfield) [inline examples/custom_target.rs:60]:
```suggestion
        // A wrapper is needed, because the `Sender` itself doesn't implement `std::io::Write`.
```

File: src/lib.rs
Comment (SirWindfield) [inline src/lib.rs:714]:
```suggestion
    /// The custom pipe can be used to send the log messages to a custom sink (for example a file).
    /// Do note that direct writes to a file can become a bottleneck due to IO operation times.
```

I don't think a notice is needed to be honest, but if we add one I'd go for something similar to the above without going into implementation details.

**Review by SirWindfield — CHANGES_REQUESTED** [review]
The rest looks decent. Maybe @jplatte could look over again as he was the initial reviewer, and a second pair of eyes is always good ;)

That being said, sorry that it took me that long.

File: src/fmt/writer/termcolor/shim_impl.rs
Comment (TilCreator) [reply src/fmt/writer/termcolor/shim_impl.rs:67]: Ah fluff, I applied the suggested changes backwards and now I can't apply this one directly. (I'm a bit new to reviews)

**Comment by TilCreator** [comment]
thx @SirWindfield (also for fixing my typos ^^)

File: examples/custom_target.rs
Comment (TilCreator) [reply examples/custom_target.rs:19]: I would agree that this should be changed, but all other examples are very similar and use this line. So a bulk change of all the examples would be better. Should I still change it?

**Comment by jplatte** [comment]
I can have a second look, sure. Probably going to take a while for me as well.

**Comment by jose-acevedoflores** [comment]
I've been playing around with this pr since I need support for logging to a file and this works great for my use case!

I did have one question about the possibility of having multiple targets like stdout and a file. Is this something out of scope for the goals of this library or is it something that could be entertained ?

**Comment by TilCreator** [comment]
(I'm just a contributor, not a maintainer)
I think this is possibly better implemented by the user, it sounds like a quite specific use case. (Maybe this could be added to the [example](examples/custom_target.rs) (and I'm still not entirely happy with the example)?)

File: src/fmt/writer/mod.rs
Comment (jplatte) [inline src/fmt/writer/mod.rs:119]: Why split the target into two fields and introduce a new `TargetType` type? I think `target` can just be wrapped in `Option` for the `build` method to work (where you can then use `self.target.take().unwrap_or_default()` to get the actual value).

File: src/fmt/writer/termcolor/extern_impl.rs
Comment (jplatte) [inline src/fmt/writer/termcolor/extern_impl.rs:75]: Can you explain why `Arc<Mutex<>>` is needed?

File: src/fmt/writer/termcolor/shim_impl.rs
Comment (jplatte) [inline src/fmt/writer/termcolor/shim_impl.rs:40]: Maybe put back the empty line before the final expression of the function.

File: src/fmt/writer/mod.rs
Comment (TilCreator) [reply src/fmt/writer/mod.rs:119]: I did try that, because I would like it way more that way, too. But the Target gets passed (and cloned) everywhere (for color checking and the `shim_impl`) that is why `TargetType` needs `Clone, Copy, Debug, Eq, Hash, PartialEq` implemented which is hard to impossible with `dyn io::Write + Send + 'static`.

File: src/fmt/writer/termcolor/extern_impl.rs
Comment (TilCreator) [reply src/fmt/writer/termcolor/extern_impl.rs:75]: Because the pipe is still partly owned by the `Builder` (it doesn't get deconstructed on `build()`).

File: src/fmt/writer/termcolor/shim_impl.rs
Comment (TilCreator) [reply src/fmt/writer/termcolor/shim_impl.rs:40]: That one just got lost ^^

**Comment by jplatte** [comment]
I'm still planning to have a look at this again, hopefully this or next week.

**Comment by jplatte** [comment]
Merged in #204.

**Pull request closed without merging**
